### PR TITLE
feat(models): Implement Pydantic models with metadata fields (fixes #3)

### LINKED ISSUE #3: [Phase 1] Task 1.3: Pydantic Models Implementation
## Description
Implement Pydantic models that match the OpenSpec schema for runtime validation.

## Dependencies
- Task 1.2: OpenSpec Schema Definition

## Acceptance Criteria
- `Section` model with recursive `sections` field
- `Document` model with all metadata fields
- Models support forward references for recursion
- Models generate stable IDs (UUID) when not provided
- Models validate against OpenSpec constraints
- Models handle Hebrew text correctly (UTF-8)
- Models provide clear validation error messages

## Tests
- Unit test: Create Section with all fields
- Unit test: Create Section with minimal fields (id only)
- Unit test: Create nested sections (3+ levels deep)
- Unit test: Auto-generate UUID when id not provided
- Unit test: Reject invalid data types
- Unit test: Reject missing required fields with clear errors
- Unit test: Handle Hebrew content correctly
- Integration test: Load YAML file into Pydantic models
- Integration test: Round-trip: Pydantic → YAML → Pydantic

## Files to Create
- `src/yaml_diffs/models/document.py`
- `src/yaml_diffs/models/section.py`
- `src/yaml_diffs/models/__init__.py`
- `tests/test_models.py`

## Phase
Phase 1: Foundation & Schema Definition

### PR DESCRIPTION
**Pull request by noamoss**
## Summary
Implements Pydantic models (`Document` and `Section`) for runtime validation of Hebrew legal documents in YAML format. The models support recursive nesting, auto-generate UUIDs, handle Hebrew text correctly, and include comprehensive metadata fields. Also creates the OpenSpec schema definition.

## Related Issues
Fixes #3

## Changes Made
- **Section Model**: Recursive structure with UUID auto-generation, Hebrew text support
- **Document Model**: Complete metadata fields (title, authors, version, source, published_date, updated_date)
- **OpenSpec Schema**: Created `legal_document_spec.yaml` with full schema definition
- **Comprehensive Tests**: 21 test cases covering all acceptance criteria (100% coverage)

## Key Features
- ✅ Recursive sections with unlimited nesting depth
- ✅ Auto-generated UUIDs when not provided
- ✅ Full Hebrew text support (UTF-8)
- ✅ Comprehensive metadata fields (title, authors, dates)
- ✅ Clear validation error messages
- ✅ OpenSpec schema alignment

## Testing
- [x] All 21 tests pass
- [x] Pre-commit hooks pass (UP045 warnings are intentional for Python 3.9 compatibility)
- [x] Code coverage: 100% for models
- [x] Linting passes (ruff)
- [x] Type checking passes (mypy)

## Checklist
- [x] Tests added/updated (21 comprehensive tests)
- [x] Documentation updated (docstrings with examples)
- [x] OpenSpec schema created
- [x] Backward compatible (all new fields optional)

## Files Changed
- `src/yaml_diffs/models/__init__.py` - Model exports
- `src/yaml_diffs/models/section.py` - Section model with recursion
- `src/yaml_diffs/models/document.py` - Document model with metadata
- `src/yaml_diffs/schema/__init__.py` - Schema package init
- `src/yaml_diffs/schema/legal_document_spec.yaml` - OpenSpec schema
- `tests/test_models.py` - Comprehensive test suite

---
🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: src/yaml_diffs/models/document.py [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+        description="List of authors, entities, or organizations. Supports Hebrew text (UTF-8).",
+    )
+    version: str = Field(
+        description="Document version identifier (e.g., '1.0', '2024-01-15').",
+    )
+    source: Optional[str] = Field(
+        default=None,
+        description="Optional source identifier or URL for the document.",
+    )
+    published_date: Optional[str] = Field(
```
Comment (noamoss) [inline src/yaml_diffs/models/document.py]: **Missing Date Format Validation**

The `published_date` and `updated_date` fields accept any string, but the OpenSpec schema specifies `format: date-time` (ISO 8601). Consider adding validation to ensure dates are in the correct format.

**Suggested fix:**
```python
from datetime import datetime
from typing import Optional

from pydantic import BaseModel, Field, field_validator

# ... in Document class ...

@field_validator('published_date', 'updated_date')
@classmethod
def validate_date_format(cls, v: Optional[str]) -> Optional[str]:
    if v is None:
        return v
    try:
        # Try parsing as ISO 8601 (supports both date and datetime)
        datetime.fromisoformat(v.replace('Z', '+00:00') if v.endswith('Z') else v)
        return v
    except ValueError:
        raise ValueError(f"Date must be in ISO 8601 format (YYYY-MM-DD or YYYY-MM-DDTHH:MM:SS), got: {v}")
```

Alternatively, use Pydantic's `datetime` type with custom serialization if you want to keep them as strings in the model but validate the format.

File: tests/test_models.py [hunk 2]
```diff
@@ -0,0 +1,433 @@
...
+        assert doc.authors == []
+
+    def test_document_rejects_missing_version(self):
+        """Test Document rejects missing required version field."""
+        with pytest.raises(ValidationError) as exc_info:
+            Document()  # version is required
+        errors = exc_info.value.errors()
+        assert len(errors) > 0
+        error_msg = str(exc_info.value).lower()
+        assert "version" in error_msg
```
Comment (noamoss) [inline tests/test_models.py]: **Missing Test for Invalid Date Formats**

Consider adding a test to verify that invalid date formats are rejected (if date validation is added) or at least document the current behavior that any string is accepted.

**Suggested addition:**
```python
def test_document_rejects_invalid_date_format(self):
    """Test Document rejects invalid date formats."""
    # If validation is added, this should raise ValidationError
    # For now, this documents current behavior
    doc = Document(version="1.0", published_date="not-a-date")
    assert doc.published_date == "not-a-date"  # Currently accepts any string
```

This test would need to be updated if date validation is implemented.

File: src/yaml_diffs/models/document.py [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+        default=None,
+        description="Publication date in ISO 8601 format (YYYY-MM-DD or YYYY-MM-DDTHH:MM:SS).",
+    )
+    updated_date: Optional[str] = Field(
+        default=None,
+        description="Last update date in ISO 8601 format (YYYY-MM-DD or YYYY-MM-DDTHH:MM:SS).",
+    )
+    sections: list[Section] = Field(
```
Comment (noamoss) [inline src/yaml_diffs/models/document.py]: **Type Consistency Note**

Line 60 uses `list[Section]` which also requires the `__future__` import or `List[Section]` from typing for Python 3.9 compatibility. This should be addressed along with the `authors` field fix.

**Review by noamoss — COMMENTED** [review]
## Code Review Summary

### Issues Found

1. **Python 3.9 Compatibility** (`document.py:41, 60`)
   - `list[str]` syntax requires `from __future__ import annotations` or `List[str]` from typing
   - Inconsistent with `section.py` which uses the `__future__` import

2. **Missing Date Validation** (`document.py:52, 56`)
   - Date fields accept any string, but OpenSpec schema specifies ISO 8601 format
   - Consider adding `@field_validator` to ensure proper date format

3. **OpenSpec Schema Mismatch** (`legal_document_spec.yaml:20-22`)
   - Schema declares `id` and `content` as required, but Pydantic model has defaults
   - Should align schema with implementation (remove from required list)

4. **Missing Test Coverage** (`test_models.py`)
   - No test for invalid date format handling
   - Consider adding test to document current behavior or validate dates

### Positive Aspects

✅ **Excellent Test Coverage**: Comprehensive test suite with 21 tests covering:
   - Model creation with various field combinations
   - Deep nesting (4+ levels)
   - Hebrew text handling
   - UUID auto-generation
   - Integration tests for YAML round-trips
   - Edge cases

✅ **Well-Structured Models**: 
   - Clear separation of concerns (Document vs Section)
   - Good use of Pydantic features (default_factory, Field descriptions)
   - Proper recursive structure implementation

✅ **Good Documentation**:
   - Comprehensive docstrings with examples
   - Clear field descriptions
   - Hebrew text support well-documented

✅ **OpenSpec Schema**: 
   - Well-defined schema definition
   - Good alignment with model structure (aside from required fields)

✅ **Code Quality**:
   - Follows project conventions
   - Proper use of type hints
   - Good error handling in tests

### Recommendations

1. Add `from __future__ import annotations` to `document.py` for Python 3.9 compatibility
2. Consider adding date format validation using Pydantic validators
3. Update OpenSpec schema to match implementation (remove `id` and `content` from required)
4. Add test for invalid date format handling

Overall, this is a solid implementation with excellent test coverage. The issues are minor and easily addressable.

**Comment by noamoss** [comment]
## 📋 Review Feedback Addressed

I've applied the following changes based on code review:

### ✅ Changes Made:

1. **Python 3.9 Compatibility** ([comment #2554287099](https://github.com/noamoss/yaml_diffs/pull/28#discussion_r2554287099))
   - Added `from __future__ import annotations` to `document.py` for Python 3.9 compatibility
   - Now consistent with `section.py` approach
   - Fixes type annotations for `list[str]` and `list[Section]`

2. **Date Format Validation** ([comment #2554287123](https://github.com/noamoss/yaml_diffs/pull/28#discussion_r2554287123))
   - Added `@field_validator` for `published_date` and `updated_date` fields
   - Validates ISO 8601 format (supports both date and datetime formats)
   - Handles 'Z' timezone indicator correctly
   - Provides clear error messages for invalid formats

3. **OpenSpec Schema Alignment** ([comment #2554287169](https://github.com/noamoss/yaml_diffs/pull/28#discussion_r2554287169))
   - Removed `id` and `content` from required fields in schema
   - Schema now matches implementation (both fields have defaults)
   - Aligns with Pydantic model behavior

4. **Test Coverage** ([comment #2554287193](https://github.com/noamoss/yaml_diffs/pull/28#discussion_r2554287193))
   - Added `test_document_rejects_invalid_date_format` to verify date validation
   - Added `test_document_accepts_valid_date_formats` to test various ISO 8601 formats
   - Tests now cover both valid and invalid date scenarios

### ✅ Validation:
- All 23 tests passed (added 2 new tests)
- Code formatting verified
- Linting passes (ruff)
- Type checking passes (mypy)
- Changes committed locally

### 📦 Next Step:
Pushing changes to remote branch...

---
Ready for re-review! 🚀

**Comment by noamoss** [comment]
## ✅ Merge Conflicts Resolved

Resolved merge conflicts with main branch:

### Conflicts Resolved:

1. **schema/__init__.py**
   - Kept main branch version with utility functions (load_schema, get_schema_version)
   - These functions are useful for schema loading and version checking

2. **schema/legal_document_spec.yaml**
   - Kept our OpenAPI 3.0.0 schema that matches our Pydantic models
   - Note: Main branch has a different JSON Schema format (from PR #25)
   - Our schema aligns with the Pydantic models implemented in this PR

### Validation:
- ✅ All 23 tests pass
- ✅ Rebased on latest main branch
- ✅ Conflicts resolved
- ✅ Ready for merge

The PR should now be mergeable! 🚀

**Comment by noamoss** [comment]
## ✅ CI/CD Failures Fixed

Fixed CI/CD test failures by reverting schema to JSON Schema format:

### Issue:
- Tests expected JSON Schema format (from PR #25)
- We had replaced it with OpenAPI 3.0.0 format
- 13 schema validation tests were failing

### Fix:
- Reverted `legal_document_spec.yaml` to JSON Schema format from main
- Updated `get_schema_version()` to support both formats (JSON Schema and OpenAPI)
- All tests now pass ✅

### Note:
The Pydantic models use a different structure than the JSON Schema. This is expected for now - the models work correctly for runtime validation. A follow-up issue can be created to align the models with the JSON Schema structure if needed.

### Validation:
- ✅ All 55 tests pass (23 model tests + 25 schema tests + 7 structure tests)
- ✅ Schema validation tests pass
- ✅ Model tests pass
- ✅ Ready for CI/CD

CI should now pass! 🚀

**Comment by noamoss** [comment]
## ✅ Fixed: Models Now Align with JSON Schema

I've fixed the mismatch between Pydantic models and the JSON Schema. The models now **fully match** the schema specification.

### What Changed

**Section Model:**
- ✅ uid=501(noam) gid=20(staff) groups=20(staff),12(everyone),61(localaccounts),79(_appserverusr),80(admin),81(_appserveradm),98(_lpadmin),701(com.apple.sharepoint.group.1),33(_appstore),100(_lpoperator),204(_developer),250(_analyticsusers),395(com.apple.access_ftp),398(com.apple.access_screensharing),399(com.apple.access_ssh),400(com.apple.access_remote_ae): Changed from  to  (matches schema pattern )
- ✅ Auto-generates UUID string when not provided
- ✅ Validates ID pattern

**Document Model:**
- ✅ Added required uid=501(noam) gid=20(staff) groups=20(staff),12(everyone),61(localaccounts),79(_appserverusr),80(admin),81(_appserveradm),98(_lpadmin),701(com.apple.sharepoint.group.1),33(_appstore),100(_lpoperator),204(_developer),250(_analyticsusers),395(com.apple.access_ftp),398(com.apple.access_screensharing),399(com.apple.access_ssh),400(com.apple.access_remote_ae) field (string)
- ✅ Made  required (was optional)
- ✅ Added required  field (DocumentType enum: law, regulation, directive, circular, policy, other)
- ✅ Added required  field (Literal["hebrew"], defaults to "hebrew")
- ✅ : Changed from  to  object with  (required) and  (optional)
- ✅ : Changed from  to  object with  (required) and  (required, ISO 8601)
- ✅ Kept optional metadata: , , 

**New Models:**
- ✅  enum (matches schema enum values)
- ✅  model (object with number + optional description)
- ✅  model (object with url + fetched_at)

**Schema Updates:**
- ✅ Added optional  field to JSON Schema
- ✅ Added optional  field to JSON Schema
- ✅ Added optional  field to JSON Schema

**Tests:**
- ✅ Updated all 26 model tests to match new structure
- ✅ All tests pass (58 total including schema validation)
- ✅ Integration tests verify YAML round-trip works correctly

**Examples:**
- ✅  loads correctly
- ✅  loads correctly

### Validation

- ✅ All 58 tests pass
- ✅ Linting passes (ruff)
- ✅ Type checking passes (mypy)
- ✅ Examples validate correctly
- ✅ Pre-commit hooks pass

### Breaking Changes

This is a **BREAKING CHANGE** - the model structure has changed significantly:
-  is now  instead of 
-  requires uid=501(noam) gid=20(staff) groups=20(staff),12(everyone),61(localaccounts),79(_appserverusr),80(admin),81(_appserveradm),98(_lpadmin),701(com.apple.sharepoint.group.1),33(_appstore),100(_lpoperator),204(_developer),250(_analyticsusers),395(com.apple.access_ftp),398(com.apple.access_screensharing),399(com.apple.access_ssh),400(com.apple.access_remote_ae), , , ,  (object),  (object)
-  is now a  object, not a string
-  is now a  object, not an optional string

The models now **correctly validate** against the JSON Schema, and examples load as expected! 🎉

File: src/yaml_diffs/models/document.py [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+    @field_validator("published_date", "updated_date")
+    @classmethod
+    def validate_date_format(cls, v: Optional[str]) -> Optional[str]:
+        """Validate that date strings are in ISO 8601 format."""
+        if v is None:
+            return v
+        try:
+            # Try parsing as ISO 8601 (supports both date and datetime)
+            # Handle 'Z' timezone indicator - only replace trailing 'Z'
+            date_str = v[:-1] + "+00:00" if v.endswith("Z") else v
```
Comment (noamoss) [inline src/yaml_diffs/models/document.py]: **Potential Edge Case in Date Validation**

The date validation logic on line 154 has a potential edge case. If `v` is a single character "Z", then `v[:-1]` would be an empty string, resulting in `"+00:00"` which is invalid.

While this is unlikely in practice, consider adding a length check:

```python
@field_validator("published_date", "updated_date")
@classmethod
def validate_date_format(cls, v: Optional[str]) -> Optional[str]:
    """Validate that date strings are in ISO 8601 format."""
    if v is None:
        return v
    try:
        # Handle 'Z' timezone indicator - only replace trailing 'Z'
        if v.endswith("Z") and len(v) > 1:
            date_str = v[:-1] + "+00:00"
        else:
            date_str = v
        datetime.fromisoformat(date_str)
        return v
    except ValueError as err:
        raise ValueError(
            f"Date must be in ISO 8601 format (YYYY-MM-DD or YYYY-MM-DDTHH:MM:SS), got: {v}"
        ) from err
```

Alternatively, you could use a regex to check for valid ISO 8601 formats before parsing.

File: src/yaml_diffs/models/document.py [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+    def validate_fetched_at_format(cls, v: str) -> str:
+        """Validate that fetched_at is in ISO 8601 format."""
+        try:
+            # Handle 'Z' timezone indicator - only replace trailing 'Z'
+            date_str = v[:-1] + "+00:00" if v.endswith("Z") else v
+            datetime.fromisoformat(date_str)
+            return v
+        except ValueError as err:
+            raise ValueError(
+                f"fetched_at must be in ISO 8601 format (YYYY-MM-DDTHH:MM:SS), got: {v}"
```
Comment (noamoss) [inline src/yaml_diffs/models/document.py]: **Inconsistent Date Format Requirements**

The `fetched_at` validator error message (line 70) says it must be in format "YYYY-MM-DDTHH:MM:SS", but the `published_date`/`updated_date` validator (line 159) says it supports both "YYYY-MM-DD" and "YYYY-MM-DDTHH:MM:SS". 

Since `fetched_at` uses the same `datetime.fromisoformat()` logic, it should also support date-only format. Consider updating the error message for consistency:

```python
raise ValueError(
    f"fetched_at must be in ISO 8601 format (YYYY-MM-DD or YYYY-MM-DDTHH:MM:SS), got: {v}"
) from err
```

Or, if `fetched_at` should always be a full timestamp (not just a date), add validation to enforce that requirement.

File: src/yaml_diffs/models/document.py [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+
+class Source(BaseModel):
+    """Document source information.
+
+    Attributes:
+        url: Original source URL (must be valid URI).
+        fetched_at: ISO 8601 timestamp when document was fetched.
+    """
+
+    url: str = Field(
```
Comment (noamoss) [inline src/yaml_diffs/models/document.py]: **Missing URL Validation**

The `Source` model's docstring says "url: Original source URL (must be valid URI)" and the schema specifies `format: uri`, but there's no Pydantic validation to ensure the URL is actually a valid URI.

Consider adding a validator:

```python
from urllib.parse import urlparse

@field_validator("url")
@classmethod
def validate_url(cls, v: str) -> str:
    """Validate that url is a valid URI."""
    try:
        result = urlparse(v)
        if not all([result.scheme, result.netloc]):
            raise ValueError(f"url must be a valid URI with scheme and netloc, got: {v}")
        return v
    except Exception as err:
        raise ValueError(f"url must be a valid URI, got: {v}") from err
```

Alternatively, you could use a library like `validators` or `pydantic.HttpUrl` type, though `HttpUrl` requires `http://` or `https://` schemes.

**Review by noamoss — COMMENTED** [review]
## Final Code Review - Excellent Work! ✅

All previous review feedback has been addressed. This is a **high-quality implementation** ready for merge.

### ✅ All Previous Issues Resolved

1. ✅ **URL Validation**: Added comprehensive URL validation with proper error handling
2. ✅ **Date Validation Edge Cases**: Fixed with length check for "Z" timezone indicator  
3. ✅ **Date Format Messages**: Aligned error messages across all date fields
4. ✅ **ID Pattern Documentation**: Clear documentation explaining UUID compatibility
5. ✅ **Test Coverage**: Comprehensive tests for URL validation added

### Code Quality Highlights

✅ **Robust Validation**:
   - URL validation with scheme and netloc checks
   - Date format validation with edge case handling
   - ID pattern validation with clear error messages
   - All validators properly handle edge cases

✅ **Excellent Test Coverage**:
   - Tests for invalid URL formats (no scheme, no netloc, invalid format)
   - Tests for date validation edge cases
   - Comprehensive integration tests
   - Edge case testing (empty strings, long text, deep nesting)

✅ **Clean Code Structure**:
   - Well-organized models with clear separation of concerns
   - Good use of Pydantic features (validators, Field descriptions)
   - Consistent error handling patterns
   - Clear documentation with examples

✅ **Schema Alignment**:
   - Models match JSON Schema specification
   - Proper handling of required vs optional fields
   - Good documentation explaining auto-generation behavior

### Summary

This PR demonstrates:
- ✅ Strong understanding of Pydantic best practices
- ✅ Attention to edge cases and error handling
- ✅ Excellent test coverage
- ✅ Good documentation
- ✅ Responsive to code review feedback

**The implementation is production-ready and follows best practices. Great work!** 🚀
